Add namespace capability to store modules

LMS007 opened the pull request:
Each module can now opt into being namespaced by exporting the value for "namespace". The `links` module  is the only module that has been namespaced.

**When using a namespace a few things change:**

- Reducers need not be aware of the namespace. The scope will be set when they are called.
- Namespaced selectors of the module will need to use the namespace because they will be given the root of the store.
- A namespaced module's state keys no longer need to be unique since it can't collide with other state keys.
- A namespaced module's state can return an array rather than an object if the state is a simple list of thing.
- There is a new helper method called getRootState which returns the root of the store which has access to both "base" and all other new namespaces.
- getRootState returns the root of the store, 

**Some behavior changes for Legacy (non-namespaced) modules:**

- Legacy modules are nested in a pseudo namespace called "base" which will be depreciated once all module are namespaced. 
- getState returns the "base" state
- Selectors of legacy modules will not need to use the namespace because they will be given a scoped state of the "base" namespace

The long term plan is to export a namespace for each module at which point we'll do a little deprecation work inside of create-store. 

_Note, this work does not namespace actions or selectors functions themselves -- that work may come later._

#1235

LMS007 commented on src/sidebar/store/util.js:
```diff
@@ -20,42 +20,62 @@ function actionTypes(updateFns) {
+function bindSelectors(namespaces, getState) {
+  const totalSelectors = {};
+  Object.keys(namespaces).forEach(namespace => {
+    const selectors = namespaces[namespace].selectors;
```
todo: add test coverage for this

LMS007 commented on src/sidebar/store/util.js:
```diff
@@ -20,42 +20,62 @@ function actionTypes(updateFns) {
+    // Some states return an array rather than an object. They need to be
+    // handled differently so we don't cast them to an object.
+    if (Array.isArray(state)) {
+      return [...fns[0](state, action)];
```
todo: add test coverage here

LMS007 replied:
The array will go away once we move all modules to namespaces. Array types are only used for namespaces modules and they can only have 1 item in the array. Using `[0]` is okay here for now until we remove the array altogether.

LMS007 commented on src/sidebar/store/modules/annotations.js:
```diff
@@ -254,15 +254,15 @@ function addAnnotations(annotations, now) {
   return function(dispatch, getState) {
     const added = annotations.filter(function(annot) {
-      return !findByID(getState().annotations, annot.id);
+      return !findByID(getState().base.annotations, annot.id);
```
thunk actions are not scoped so they need to drill into the "base" namespace until the module is converted.

LMS007 commented on src/sidebar/store/util.js:
```diff
@@ -20,42 +20,62 @@ function actionTypes(updateFns) {
+    Object.keys(selectors).forEach(selector => {
+      totalSelectors[selector] = function() {
+        const args = [].slice.apply(arguments);
+        if (scopeSelector) {
```
when using scopeSelector, the selector does not need to drill into the top level namespace scope. This is a way to allow legacy modules to work without changes but keep them under the "base" namespace.

Only the legacy (non-namespaced) modules get `scopeSelector=true`. This will eventually be deprecated when all modules are switched over.

robertknight replied:
We'll still need to come up with something for selectors that access state from multiple modules, or some equivalent for what those selectors do.

LMS007 replied:
namespaced selectors get the root store so thy will still have access to everything , but paths will change of course.

robertknight commented on src/sidebar/store/test/util-test.js:
```diff
@@ -36,6 +51,15 @@ describe('reducer utils', function() {
   });
 
   describe('#createReducer', function() {
+    it('can not return undefined during initialization', function() {
```
It wasn't immediately clear to me what this was trying to test. `it("returns an object if input state is undefined and action is unrecognized")` might be a more direct description, with a brief note on when this happens as a comment below.

LMS007 replied:
The action does not matter here, but perhaps using "fake" is misleading for that action

robertknight commented on src/sidebar/store/test/util-test.js:
```diff
@@ -109,19 +133,19 @@ describe('reducer utils', function() {
+      });
+      const bound = util.bindSelectors(fixtures.selectors, getState);
+      // test namespaced selector
+      assert.equal(bound.countAnnotations1(), 1);
```
`countAnnotations1` is a non-namespaced selector rather than a namespaced one. Conversely `countAnnotations2` is a namespaced selector.

LMS007 replied:
good catch!

robertknight commented on src/sidebar/store/util.js:
```diff
@@ -20,42 +20,62 @@ function actionTypes(updateFns) {
     if (!fns) {
       return state;
     }
+    // Some states return an array rather than an object. They need to be
```
By "some states return an array" did you mean "some modules use an array as their state"?

LMS007 replied:
yes

robertknight commented on src/sidebar/store/util.js:
```diff
@@ -20,42 +20,62 @@ function actionTypes(updateFns) {
-    return bound;
-  }, {});
+function bindSelectors(namespaces, getState) {
+  const totalSelectors = {};
```
I think `allSelectors` or `combinedSelectors` would be a better name than `totalSelectors` here, since that implies a numeric sum to me.

LMS007 replied:
I will make that change

robertknight commented on src/sidebar/store/util.js:
```diff
@@ -20,42 +20,62 @@ function actionTypes(updateFns) {
-  createReducer: createReducer,
+  actionTypes,
+  bindSelectors,
+  createReducer,
```
👍 - Use of object shorthand is preferred in new code.

robertknight commented on src/sidebar/store/create-store.js:
```diff
@@ -26,28 +29,86 @@ const { createReducer, bindSelectors } = require('./util');
+  const store = configureStore({
+    reducer: totalReducers,
+    middleware: [thunk, ...middleware],
+    devTools: process.env.NODE_ENV !== 'production',
```
I was about to write that there wasn't much point enabling the devtools as it didn't support applications that run in iframes. It looks like that [might have changed](https://github.com/zalmoxisus/redux-devtools-extension/issues/50) though. I'm happy to add this if it works and is useful, but see caveat about redux-starter-kit below.

LMS007 commented:
- Updated this to have test full coverage again.
- Fixed the issue with link module (I originally slotted that for the first namespaced example, then verted it back, but in that effort, I didn't change a few peripheral lines that causes some errors. 
- `links` module is now namespaced. 
- redux-start-kit is removed and this uses combineReducers directly.

robertknight commented on src/sidebar/store/test/util-test.js:
```diff
@@ -36,6 +51,16 @@ describe('reducer utils', function() {
+      // See redux.js:assertReducerShape in the "redux" package.
+      const reducer = util.createReducer(fixtures.update);
+      const initialState = reducer(undefined, {
+        type: 'ADD_ANNOTATIONS',
```
I think it might be clearer here to name this `DUMMY_ACTION` or something to indicate that it is an action that the reducer doesn't handle. Internally Redux will dispatch actions with [randomized "type" strings](https://github.com/reduxjs/redux/blob/9c9a4d2a1c62c9dbddcbb05488f8bd77d24c81de/src/utils/actionTypes.js#L16) for various purposes.

robertknight commented on src/sidebar/store/test/util-test.js:
```diff
@@ -105,23 +130,39 @@ describe('reducer utils', function() {
       assert.deepEqual(newState, { firstCounter: 6, secondCounter: 11 });
     });
+
+    it('creates a reducer that can consume a state that returns an array', function() {
```
```suggestion
    it('supports reducer functions that return an array', function() {
```

I had trouble parsing this as "a state" is just an object, it doesn't return anything.

LMS007 replied:
better :)

LMS007 commented on src/sidebar/store/create-store.js:
```diff
@@ -26,28 +26,94 @@ const { createReducer, bindSelectors } = require('./util');
+      initialState[module.namespace] = module.init(...initArgs);
+      allReducers[module.namespace] = createReducer(module.update);
+      allSelectors[module.namespace] = {
+        selectors: module.selectors,
```
@robertknight, I intended useLocalState to be a property of the intermediary translation object allSelectors, but not part of any modules (since its temporary). In other words, we would never export that value from any module. It's just hard coded to the "base" set but not any other namespaced module. Do you think that is a issue?

My intention here is that all namespaced module selectors get the top level root state parameter  and would be responsible for drilling down into each namespace they want to access. This allows module selectors to see other states too.

Thoughts on that?
